src/OpenTelemetry.Sampler.AWS/AWSXRayRemoteSampler.cs

```diff
@@ -25,38 +28,70 @@ namespace OpenTelemetry.Sampler.AWS;
+    internal AWSXRayRemoteSampler(Resource resource, TimeSpan pollingInterval, string endpoint, Clock clock)
     {
+        this.Resource = resource;
         this.PollingInterval = pollingInterval;
         this.Endpoint = endpoint;
-        this.Client = new AWSXRaySamplerClient(endpoint);
+        this.Clock = clock;
+        this.ClientId = GenerateClientId();
+        this.Client = new AWSXRaySamplerClient(this.Endpoint);
+        this.FallbackSampler = new FallbackSampler(this.Clock);
+        this.RulesCache = new RulesCache(this.Clock, this.ClientId, this.Resource, this.FallbackSampler);
+
+        // upto 5 seconds of jitter for rule polling
+        this.RulePollerJitter = TimeSpan.FromMilliseconds(Random.Next(1, 5000));
 
-        // execute the first update right away
-        this.RulePollerTimer = new Timer(this.GetAndUpdateSampler, null, TimeSpan.Zero, this.PollingInterval);
+        // execute the first update right away and schedule subsequent update later.
+        this.RulePollerTimer = new Timer(this.GetAndUpdateRules, null, TimeSpan.Zero, Timeout.InfiniteTimeSpan);
     }
 
-    internal TimeSpan PollingInterval { get; }
+    internal TimeSpan RulePollerJitter { get; set; }
+
+    internal Clock Clock { get; set; }
+
+    internal string ClientId { get; set; }
 
-    internal string Endpoint { get; }
+    internal Resource Resource { get; set; }
 
-    internal AWSXRaySamplerClient Client { get; }
+    internal string Endpoint { get; set; }
 
-    internal Timer RulePollerTimer { get; }
+    internal AWSXRaySamplerClient Client { get; set; }
+
+    internal RulesCache RulesCache { get; set; }
+
+    internal Timer RulePollerTimer { get; set; }
+
+    internal TimeSpan PollingInterval { get; set; }
+
+    internal Trace.Sampler FallbackSampler { get; set; }
 
     /// <summary>
     /// Initializes a <see cref="AWSXRayRemoteSamplerBuilder"/> for the sampler.
     /// </summary>
+    /// <param name="resource">an instance of <see cref="Resources.Resource"/>
+    /// to identify the service attributes for sampling. This resource should
+    /// be the same as what the OpenTelemetry SDK is configured with.</param>
     /// <returns>an instance of <see cref="AWSXRayRemoteSamplerBuilder"/>.</returns>
-    public static AWSXRayRemoteSamplerBuilder Builder()
+    public static AWSXRayRemoteSamplerBuilder Builder(Resource resource)
     {
-        return new AWSXRayRemoteSamplerBuilder();
+        return new AWSXRayRemoteSamplerBuilder(resource);
     }
 
     /// <inheritdoc/>
     public override SamplingResult ShouldSample(in SamplingParameters samplingParameters)
     {
-        // TODO: add the actual functionality for sampling.
-        throw new System.NotImplementedException();
+        if (this.RulesCache.Expired())
+        {
+            return this.FallbackSampler.ShouldSample(in samplingParameters);
```

Comment:
Is a warning log generated somewhere when the cache expires and we use the fallback?

This could be very useful in debugging with a customer.

Reply:
Not yet. I can add one.